lib/phoenix/endpoint/cowboy2_websocket.ex

```diff
@@ -0,0 +1,116 @@
+defmodule Phoenix.Endpoint.Cowboy2WebSocket do
+  # Implementation of the WebSocket transport for Cowboy.
+  @moduledoc false
+
+  @behaviour :cowboy_websocket
+  @connection Plug.Adapters.Cowboy2.Conn
+  @already_sent {:plug_conn, :sent}
+
+  def init(req, {module, opts}) do
+    # Is this too brittle?
+    conn = @connection.conn(req)
+    case module.init(conn, opts) do
+      {:ok, _, _} ->
+        {_, socket, _} = opts
+        %{websocket: {_, config}} = socket.__transports__
+        timeout = Keyword.fetch!(config, :timeout)
+        {:cowboy_websocket, req, {module, req, opts}, %{idle_timeout: timeout}}
+      {:error, %{adapter: {@connection, req}}} ->
+        {:error, req}
+    end
+  end
+
+  def resume(module, fun, args) do
+    try do
+      apply(module, fun, args)
+    catch
+      kind, [{:reason, reason}, {:mfa, _mfa}, {:stacktrace, stack} | _rest] ->
+        reason = format_reason(kind, reason, stack)
+        exit({reason, {__MODULE__, :resume, []}})
+    else
+      {:suspend, module, fun, args} ->
+        {:suspend, __MODULE__, :resume, [module, fun, args]}
+      _ ->
+        # We are forcing a shutdown exit because we want to make
+        # sure all transports exits with reason shutdown to guarantee
+        # all channels are closed.
+        exit(:shutdown)
+    end
+  end
+
+  defp format_reason(:exit, reason, _), do: reason
+  defp format_reason(:throw, reason, stack), do: {{:nocatch, reason}, stack}
+  defp format_reason(:error, reason, stack), do: {reason, stack}
+
+  ## Websocket callbacks
+
+  def websocket_init({module, req, opts}) do
+    conn = @connection.conn(req)
```

Comment:
Why do we have the same logic in websocket_init and init?

Reply:
And same logic I mean the module.init call.

Reply:
So, we need to do the first inside init here (force SSL, origin check, etc.) - this is the HTTP part of the request (before the upgrade).

We then want to do the Transport.connect part of the code here.

https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/transports/websocket.ex#L77

Perhaps we should split the logic of init in the transport and update the Cowboy 1 adapter to do both?

Reply:
@Gazler yes, we should split the logic somehow. I have been planning to rework this part for almost 2 years now so I am glad I have procrastinated. :P Feel free to break it apart so it works on both, later I will come back and revisit everything. :)

Reply:
@josevalim Turns out splitting this was way trickier than I first anticipated. I pushed a different fix though. Could you check it out please?